lib/smf/smf.c

```diff
@@ -62,28 +62,63 @@ static const struct smf_state *get_child_of(const struct smf_state *states,
 	return NULL;
 }
 
-static const struct smf_state *get_last_of(const struct smf_state *states)
+__unused static const struct smf_state *get_last_of(const struct smf_state *states)
 {
 	return get_child_of(states, NULL);
 }
 
 /**
- * @brief Execute all ancestor entry actions
+ * @brief Find the Least Common Ancestor (LCA) of two states
+ *
+ * @param source transition source
+ * @param dest transition destination
+ * @return LCA state, or NULL if states have no LCA.
+ */
+__unused static const struct smf_state *get_lca_of(const struct smf_state *source,
+						   const struct smf_state *dest)
+{
+	for (const struct smf_state *ancestor = source->parent; ancestor != NULL;
+	     ancestor = ancestor->parent) {
+		if (share_paren(dest, ancestor)) {
+			if (ancestor == dest) {
+				/* share_paren get confused if dest is a parent of source */
+				return ancestor->parent;
+			}
+			return ancestor;
```

Comment:
I think you can just skip the call to `share_paren()` when the ancestor is the same as the destination.
```
if (ancestor == dest) {
    return ancestor->parent;
} else if (sharen_paren(dest, ancestor)) {
    return ancestor;
```